### ISSUE TITLE
Bug: [no-base-to-string] Should detect @@toPrimitive and valueOf

### ISSUE THREAD

**Issue by Josh-Cena** [typescript-eslint/typescript-eslint#6048]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=4.8.4&sourceType=module&code=CYUwxgNghgTiAEYD2A7AzgF3lAXPA3gFDzwDaAygJ4C2ARkhAHQZIAKMAltRxhwG4gAugAoAlHkycUAcwDchAL7zCAAwAk+KApXygA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oCMBDZRLXxdk+KkwDm6KImjQO0SODABfECqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
declare const a: {
  [Symbol.toPrimitive](): string;
};

`${a}`;
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-base-to-string": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

This should not be an error; `a` has a `@@toPrimitive` method and it even explicitly returns a string.

### Actual Result

TS-ESLint believes it returns `[object Object]`.

### Additional Info

As soon as you have a `@@toPrimitive` method on an object, it always takes priority during type coercion. See:

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toPrimitive
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#type_coercion

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.43.0` |
| `@typescript-eslint/parser` | `5.43.0` |
| `TypeScript` | `4.8.4` |
| `ESLint` | `8.15.0` |
| `node` | `web` |

**Comment by Josh-Cena** [comment]
Additionally, it may be a good idea to detect `valueOf` as well when concatenating using `+`, because that calls `valueOf` in priority. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Addition

```js
console.log({ valueOf: () => 1 } + "");
// ESLint error: '{ valueOf: () => 1 }' will evaluate to '[object Object]' when stringified.
// Actual result: "1"
```

But that sounds more niche: people usually override `toString` before they think about `valueOf`. In any case, if anyone approaches this issue, it should be fairly trivial to extend it to `valueOf`.

**Comment by ronami** [comment]
@Josh-Cena from what I can see, a similar fix may be necessary for `@typescript-eslint/restrict-template-expressions` and `@typescript-eslint/restrict-plus-operands` too ([see here on the playground](https://typescript-eslint.io/play/#ts=4.8.4&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3lAXPA3gFDzwDaAygJ4C2ARkhAHQZIAKMAltRxhwG4gAugAoAlHkycUAcwDchAL7zCAAwAk%2BKApXLk6LLXgBebPADU8AOSX5QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oCMBDZRLXxdk%2BKkwDm6KImjQO0SABpw2PEVLIK1Og2ZtoKMZTJ18iALbF4vc-QAexQ8mSUOTVBkiz5ildigCEnIqGnpkRhZWZ2NTWmtYZFoOUmheJgATTxk5BUhVAF8QAqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)).

Does this make sense? I'll be happy to push a PR to address this.

I think the case of `valueOf` may be a bit more complex. Objects overriding `valueOf` (like the example above) can be used for expressions with the plus operand but can't necessarily be used in a template expression. Even once it's fixed on typescript-eslint's rules, it may still appear as an error for other rules (like `prefer-template`, which will auto-fix a valid usage into an invalid one).

**Comment by Josh-Cena** [comment]
I'm not sure if those need fixing? They maintain an allowlist of types so even if a type has a custom conversion method it won't be automatically allowed by default. The only rule whose purpose is to prevent you from calling `Object.prototype.toString` is this rule, and I'm proposing that we fully encode the type conversion process to avoid falsely considering certain types as "using `Object.prototype.toString`"
